documentation/varnish_config.md
# varnish_config

[Back to resource list](../README.md#resources)

Configures the Varnish service through the systemd unit and `/etc/varnish/varnish.params`.

## Actions

| Action | Description |
| --- | --- |
| `:configure` | Creates the varnish configuration file from template |
| `:unconfigure` | Removes the files created by `:configure` |

## Properties

| Name | Type | Default |
| --- | --- | --- |
| `conf_source`              | String                  | `default_systemd.erb`               |
| `max_open_files`           | Integer                 | `131_072`                           |
| `max_locked_memory`        | Integer                 | `82_000`                            |
| `major_version`            | Minor supported release | installed major version             |
| `instance_name`            | String                  | Node hostname                       |
| `listen_address`           | String                  | `nil`                               |
| `listen_port`              | Integer                 | `6081`                              |
| `secondary_listen_address` | String                  | `nil`                               |
| `secondary_listen_port`    | Integer                 | `nil`                               |
| `admin_listen_address`     | String                  | `'127.0.0.1'`                       |
| `admin_listen_port`        | Integer                 | `6082`                              |
| `storage`                  | `'malloc'` or `'file'`  | `'file'`                            |
| `file_storage_path`        | String                  | `'/var/lib/varnish/%s_storage.bin'` |
| `file_storage_size`        | String                  | `'1G'`                              |
| `malloc_percent`           | Integer                 | `33`                                |
| `malloc_size`              | String                  | `nil`                               |
| `path_to_secret`           | String                  | `'/etc/varnish/secret'`             |
| `reload_cmd`               | String                  | version/platform default            |

You can also send a hash to `parameters` which will add additional parameters to the varnish daemon via the `-p` option. The default hash is:

```ruby
{
  'thread_pools' => '4',
  'thread_pool_min' => '5',
  'thread_pool_max' => '500',
  'thread_pool_timeout' => '300'
}
```

## Examples

```ruby
varnish_config 'default' do
  listen_address '0.0.0.0'
  listen_port 80
  storage 'malloc'
  malloc_percent 33
end
```


CHANGELOG.md
# Varnish Changelog

All notable changes to this project will be documented in this file.

Standardise files with files in sous-chefs/repo-management
Standardise files with files in sous-chefs/repo-management

## [6.0.0](https://github.com/sous-chefs/varnish/compare/v5.0.22...v6.0.0) (2026-05-07)


### ⚠ BREAKING CHANGES

* migrate varnish to custom resources ([#268](https://github.com/sous-chefs/varnish/issues/268))

### Features

* migrate varnish to custom resources ([#268](https://github.com/sous-chefs/varnish/issues/268)) ([c87bbd5](https://github.com/sous-chefs/varnish/commit/c87bbd528c3b15e46d473e561204afcfa98a733b))


### Bug Fixes

* **ci:** Update workflows to use release pipeline ([#263](https://github.com/sous-chefs/varnish/issues/263)) ([1b50e74](https://github.com/sous-chefs/varnish/commit/1b50e74f948f7d9870a8cdfed6144010c509d76b))

## [5.0.22](https://github.com/sous-chefs/varnish/compare/5.0.21...v5.0.22) (2025-10-16)


### Bug Fixes

* **ci:** Update workflows to use release pipeline ([#263](https://github.com/sous-chefs/varnish/issues/263)) ([1b50e74](https://github.com/sous-chefs/varnish/commit/1b50e74f948f7d9870a8cdfed6144010c509d76b))

## 5.0.20 - *2024-11-18*

Standardise files with files in sous-chefs/repo-management

Standardise files with files in sous-chefs/repo-management

Standardise files with files in sous-chefs/repo-management

Standardise files with files in sous-chefs/repo-management

Standardise files with files in sous-chefs/repo-management

## 5.0.12 - *2023-04-07*

Standardise files with files in sous-chefs/repo-management

## 5.0.9 - *2023-04-01*

Standardise files with files in sous-chefs/repo-management

## 5.0.8 - *2023-03-20*

Standardise files with files in sous-chefs/repo-management

## 5.0.7 - *2023-03-15*

Standardise files with files in sous-chefs/repo-management

## 5.0.3 - *2023-02-23*

Standardise files with files in sous-chefs/repo-management

## 5.0.2 - *2023-02-15*

Standardise files with files in sous-chefs/repo-management

## 5.0.1 - *2022-12-13*

Standardise files with files in sous-chefs/repo-management

Standardise files with files in sous-chefs/repo-management

* Standardise files with files in sous-chefs/repo-management
* Remove delivery folder
* Update tested platforms

## 5.0.0 - *2021-10-14*

* Enable `unified_mode`
* Require Chef >= 15.5
* Remove chef-sugar dependency
* Fix support for RHEL 8
* Add new known Varnish releases
* Fix reload when using Varnish >= 7.0
* Use apt pinning on Debian/Ubuntu when using vendor packages
* Remove sysvinit support
* Deprecate various unused parameters in `varnish_config` resource
* Various cookstyle fixes
* Update test kitchen to standard modern platforms
* Switch to InSpec for integration testing

## 4.1.2 - *2021-08-30*

* Standardise files with files in sous-chefs/repo-management

## 4.1.1 - *2021-06-01*

* resolved cookstyle error: test/fixtures/cookbooks/disable_ipv6/metadata.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/fixtures/cookbooks/disable_ipv6/recipes/default.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/fixtures/cookbooks/disable_ipv6/recipes/disable_ipv6.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/distro/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/distro/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/full_stack/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/full_stack/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish3/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish3/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish4/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish4/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish41/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish41/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish5/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish5/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish51/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish51/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish52/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish52/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish60/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish60/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish61/serverspec/default_spec.rb:1:1 convention: `Style/Encoding`
* resolved cookstyle error: test/integration/varnish61/serverspec/spec_helper.rb:1:1 convention: `Style/Encoding`

## 4.1.0 - 2020-05-05

### Added

### Changed

* Updated the yum_repository resource to use the newer baseurl property instead of the legacy url property
* Migrated to actions
* resolved cookstyle error: resources/config.rb:6:1 refactor: `ChefStyle/TrueClassFalseClassResourceProperties`
* resolved cookstyle error: resources/log.rb:4:1 refactor: `ChefStyle/TrueClassFalseClassResourceProperties`

resources/config.rb
# frozen_string_literal: true

provides :varnish_config
unified_mode true

property :conf_source, String, default: 'default_systemd.erb'
property :conf_cookbook, String, default: 'varnish'
property :conf_path, String, default: '/etc/systemd/system/varnish.service'

# Service config options
property :start_on_boot, [true, false],
          default: true,
          deprecated: 'This property has been deprecated and will be removed in a future release'
property :max_open_files, Integer, default: 131_072
property :max_locked_memory, Integer, default: 82_000
property :instance_name, String, default: lazy { VarnishCookbook::Helpers.hostname }
property :major_version, Float,
          equal_to: [3.0, 4.0, 4.1, 5, 5.0, 5.1, 5.2, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 7.0],
          default: lazy { VarnishCookbook::Helpers.installed_major_version }

# Daemon options
property :listen_address, String, default: '0.0.0.0'
property :listen_port, Integer, default: 6081
property :secondary_listen_address, [String, nil]
property :secondary_listen_port, [Integer, nil]
property :path_to_vcl, String, default: '/etc/varnish/default.vcl'
property :admin_listen_address, String, default: '127.0.0.1'
property :admin_listen_port, Integer, default: 6082
property :user, String,
          default: 'varnish',
          deprecated: 'This property has been deprecated and will be removed in a future release'
property :group, String,
          default: 'varnish',
          deprecated: 'This property has been deprecated and will be removed in a future release'
property :ccgroup, [String, nil],
          deprecated: 'This property has been deprecated and will be removed in a future release'
property :ttl, Integer,
          default: 120,
          deprecated: 'This property has been deprecated and will be removed in a future release'
property :storage, String, default: 'file', equal_to: %w(file malloc)
property :file_storage_path, String, default: '/var/lib/varnish/%s_storage.bin'
property :file_storage_size, String, default: '1GB'
property :malloc_percent, [Integer, nil], default: 33
property :malloc_size, [String, nil]
property :parameters, Hash, default:
    {
      'thread_pools' => '4',
      'thread_pool_min' => '5',
      'thread_pool_max' => '500',
      'thread_pool_timeout' => '300',
    }
property :path_to_secret, String, default: '/etc/varnish/secret'
property :reload_cmd, [String, nil]

default_action :configure

action :configure do
  extend VarnishCookbook::Helpers
  systemd_daemon_reload

  malloc_default = percent_of_total_mem(node['memory']['total'], new_resource.malloc_percent)
  reload_command = new_resource.reload_cmd || default_reload_cmd(new_resource.major_version)

  template '/etc/varnish/varnish.params' do
    action :create
    variables(
      major_version: new_resource.major_version,
      malloc_size: new_resource.malloc_size || malloc_default,
      config: new_resource,
      reload_cmd: reload_command
    )
    cookbook 'varnish'
  end

  service 'varnish' do
    supports restart: true, reload: true if new_resource.major_version < 5

    # https://github.com/varnishcache/varnish-cache/issues/2316
    supports restart: false, reload: true if new_resource.major_version >= 5

    action :nothing
  end

  # The reload-vcl script doesn't support the -j option in 4.1 and breaks reload on ubuntu.
  # This is fixed upstream but could cause issues if you are using the distro package.
  cookbook_file '/usr/share/varnish/reload-vcl' do
    extend VarnishCookbook::Helpers
    source 'reload-vcl'
    cookbook 'varnish'
    mode '0755'
    only_if { platform_family?('debian') }
  end

  execute 'generate secret file' do
    command "dd if=/dev/random of=#{new_resource.path_to_secret} count=1"
    creates new_resource.path_to_secret
    only_if { new_resource.major_version >= 6.1 }
  end

  template new_resource.conf_path do
    path new_resource.conf_path
    source new_resource.conf_source
    cookbook new_resource.conf_cookbook
    owner 'root'
    group 'root'
    mode '0644'
    variables(
      major_version: new_resource.major_version,
      malloc_size: new_resource.malloc_size || malloc_default,
      config: new_resource,
      reload_cmd: reload_command
    )
    notifies :restart, 'service[varnish]', :delayed
    notifies :run, 'execute[systemctl-daemon-reload]', :immediately
  end
end

action :unconfigure do
  extend VarnishCookbook::Helpers
  systemd_daemon_reload

  service 'varnish' do
    action :nothing
  end

  file '/etc/varnish/varnish.params' do
    action :delete
    notifies :restart, 'service[varnish]', :delayed
  end

  file '/usr/share/varnish/reload-vcl' do
    action :delete
    only_if { platform_family?('debian') }
  end

  file new_resource.path_to_secret do
    action :delete
    only_if { new_resource.major_version >= 6.1 }
  end

  file new_resource.conf_path do
    action :delete
    notifies :run, 'execute[systemctl-daemon-reload]', :immediately
  end
end
